sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,843 @@
+      if next_op is None:
+        break
+      to_visit.extend(self._get_all_subgroups_and_ops(next_op))
+      if isinstance(next_op, dsl.ParallelFor):
+        op_name_to_op[next_op.name] = next_op
+
+    return op_name_to_op
+
+  def _get_all_subgroups_and_ops(self, op):
+    """Get all ops and groups contained within this group."""
+    subgroups = []
+    if hasattr(op, 'ops'):
+      subgroups.extend(op.ops)
+    if hasattr(op, 'groups'):
+      subgroups.extend(op.groups)
+    return subgroups
+
+  def _get_inputs_outputs(
+      self,
+      pipeline,
+      root_group,
+      op_groups,
+      opsgroup_groups,
+      condition_params,
+      op_name_to_for_loop_op: Dict[Text, dsl.ParallelFor],
+      pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+  ):
+    """Get inputs and outputs of each group and op.
+
+    Returns:
+      A tuple (inputs, outputs).
+      inputs and outputs are dicts with key being the group/op names and values being list of
+      tuples (param_name, producing_op_name). producing_op_name is the name of the op that
+      produces the param. If the param is a pipeline param (no producer op), then
+      producing_op_name is None.
+    """
+    inputs = defaultdict(set)
+    outputs = defaultdict(set)
+
+    for op in pipeline.ops.values():
+      # op's inputs and all params used in conditions for that op are both considered.
+      from kfp.components.structures import ComponentSpec
+      for param in op.inputs + list(condition_params[op.name]):
+        # if the value is already provided (immediate value), then no need to expose
+        # it as input for its parent groups.
+        if param.value:
+          continue
+        if param.op_name:
+          upstream_op = pipeline.ops[param.op_name]
+          upstream_groups, downstream_groups = \
+            self._get_uncommon_ancestors(op_groups, opsgroup_groups, upstream_op, op)
+          for i, group_name in enumerate(downstream_groups):
+            if i == 0:
+              # If it is the first uncommon downstream group, then the input comes from
+              # the first uncommon upstream group.
+              inputs[group_name].add((param.full_name, upstream_groups[0]))
+            else:
+              # If not the first downstream group, then the input is passed down from
+              # its ancestor groups so the upstream group is None.
+              inputs[group_name].add((param.full_name, None))
+          for i, group_name in enumerate(upstream_groups):
+            if i == len(upstream_groups) - 1:
+              # If last upstream group, it is an operator and output comes from container.
+              outputs[group_name].add((param.full_name, None))
+            else:
+              # If not last upstream group, output value comes from one of its child.
+              outputs[group_name].add((param.full_name, upstream_groups[i+1]))
+        else:
+          if not op.is_exit_handler:
```

Comment:
I think we can raise an error when `op.is_exit_handler` because it's yet to be supported.